File: apps/chess-coach/harness/src/main/kotlin/harness/MaiaEngine.kt
package harness

/**
 * Maia/lc0-specific UCI wrapper with difficulty-based weight switching.
 * Restarts the engine when the difficulty (and thus weights file) changes.
 */
class MaiaEngine(
    private val weightsDir: String = "/app/weights",
    private val defaultTimeoutMs: Long = 30_000L
) {
    private var driver: UciDriver? = null
    private var currentDifficulty: String? = null

    val isRunning: Boolean get() = driver?.isRunning == true

    fun getMove(fen: String, difficulty: String): String {
        ensureReady(difficulty)

        val d = driver ?: throw IllegalStateException("Maia engine not started")
        d.send("position fen $fen")
        d.send("go nodes 1") // Maia is a policy network; 1 node is enough

        val lines = d.readUntil("bestmove", defaultTimeoutMs)
        for (line in lines) {
            if (line.startsWith("bestmove")) {
                return line.substringAfter("bestmove ").substringBefore(" ")
            }
        }
        throw IllegalStateException("Maia did not return a bestmove")
    }

    private fun ensureReady(difficulty: String) {
        if (driver != null && currentDifficulty == difficulty && driver!!.isRunning) return

        stop()

        val weightsFile = when (difficulty) {
            "advanced" -> "maia-1600.pb.gz"
            "expert" -> "maia-2200.pb.gz"
            else -> "maia-1100.pb.gz"
        }

        System.err.println("Starting lc0 with weights: $weightsFile")

        val d = UciDriver(
            listOf("lc0", "--weights=$weightsDir/$weightsFile", "--backend=blas"),
            defaultTimeoutMs
        )
        d.start()

        // Wait for neural network to finish loading
        d.send("isready")
        d.waitFor("readyok", defaultTimeoutMs)

        driver = d
        currentDifficulty = difficulty
    }

    fun stop() {
        driver?.stop()
        driver = null
        currentDifficulty = null
    }
}

File: projects/chess-coach/apps/server/src/MaiaEngine.ts
import { UciDriver } from "./UciDriver.ts";

/**
 * Maia/lc0 UCI wrapper with difficulty-based weight switching.
 * Mirrors the Kotlin MaiaEngine (_spec/api/behavior.md §7.1): OwnBook +
 * openings.bin, Temperature 0.5, `go nodes 1`, restart on difficulty change.
 * (No TempDecayMoves — that is the browser engine's variety mechanism.)
 */
export class MaiaEngine {
  private driver: UciDriver | null = null;
  private currentDifficulty: string | null = null;

  constructor(
    private readonly weightsDir: string = process.env.MAIA_WEIGHTS_DIR ?? "/app/weights",
    private readonly defaultTimeoutMs = 30_000,
  ) {}

  get isRunning(): boolean {
    return this.driver?.isRunning === true;
  }

  async getMove(fen: string, difficulty: string): Promise<string> {
    await this.ensureReady(difficulty);
    const d = this.driver;
    if (!d) throw new Error("Maia engine not started");

    d.send(`position fen ${fen}`);
    d.send("go nodes 1"); // Maia is a policy network; 1 node is enough

    const lines = await d.readUntil("bestmove", this.defaultTimeoutMs);
    let bestMove = "";
    for (const line of lines) {
      if (line.startsWith("bestmove")) {
        bestMove = line.split("bestmove ")[1]?.split(" ")[0]?.trim() ?? "";
      }
    }
    if (bestMove !== "") return bestMove;
    throw new Error("Maia did not return a bestmove");
  }

  private async ensureReady(difficulty: string): Promise<void> {
    if (this.driver && this.currentDifficulty === difficulty && this.driver.isRunning) return;

    this.stop();

    const weightsFile =
      difficulty === "advanced"
        ? "maia-1600.pb.gz"
        : difficulty === "expert"
          ? "maia-2200.pb.gz"
          : "maia-1100.pb.gz";

    console.error(`Starting lc0 with weights: ${weightsFile}`);

    const d = new UciDriver(
      ["lc0", `--weights=${this.weightsDir}/${weightsFile}`, "--backend=blas"],
      this.defaultTimeoutMs,
    );
    await d.start();

    d.send("setoption name OwnBook value true");
    d.send(`setoption name BookFile value ${this.weightsDir}/openings.bin`);
    d.send("setoption name Temperature value 0.5");
    d.send("isready");
    await d.waitFor("readyok", this.defaultTimeoutMs);

    this.driver = d;
    this.currentDifficulty = difficulty;
  }

  stop(): void {
    this.driver?.stop();
    this.driver = null;
    this.currentDifficulty = null;
  }
}
